### PR TITLE
Weilian/expand federation schema

### PR DESCRIPTION
**Pull request by weilianchu**
This PR allows us to get an introspection query result from a merged federation schema. This output will exactly match an introspection output, allowing users to get the full introspection query result without a single service containing the whole schema.

### CODE REVIEW

File: federation/schema.go [hunk 1]
```diff
@@ -196,14 +192,45 @@ func ConvertVersionedSchemas(schemas serviceSchemas) (*SchemaWithFederationInfo,
...
 		serviceSchemasByName[service] = serviceSchema
-
 		serviceSchemas = append(serviceSchemas, serviceSchema)
 	}
 
+	return serviceNames, serviceSchemas, serviceSchemasByName, nil
+}
+
+// MergeSchemas takes the introspection query results for all
+// verions of all executor services, and merges them into a
```
Comment (alberttjin) [inline federation/schema.go:206]: nit: versions

File: federation/merge_schemas.go [hunk 2]
```diff
@@ -312,6 +338,36 @@ func mergeEnumValues(a, b []introspectionEnumValue, mode MergeMode) ([]introspec
...
+	names := make([]string, 0, len(directives))
+	for name := range directives {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+
+	merged := make([]introspectiveDirective, 0, len(names))
+	for _, name := range names {
+		p := directives[name]
+		if len(p) == 1 {
```
Comment (alberttjin) [inline federation/merge_schemas.go:358]: noob question:

what if the mode is union and the length of p is more than 1, but the two values are different? Then wouldn't we just be taking the first value? Should we check that they are the same similar to how we merge types below?

File: federation/merge_schemas.go [hunk 3]
```diff
@@ -312,6 +338,36 @@ func mergeEnumValues(a, b []introspectionEnumValue, mode MergeMode) ([]introspec
 	return merged, nil
 }
 
+func mergeDirectives(a, b []introspectiveDirective, mode MergeMode) ([]introspectiveDirective, error) {
```
Comment (KevinXing) [inline federation/merge_schemas.go:341]: Unlike the types that can be different in gql servers, directive is a global feature that should be supported by the merged schema. 
Our directive introspection queries are hardcoded pre federation. I think it is reasonable to also hardcode them in the merged introspection schema.

Comment (weilianchu) [reply federation/merge_schemas.go:341]: Is the implication here that the directives will be the same for all federated schemas?

Comment (KevinXing) [reply federation/merge_schemas.go:341]: I think so, directives is a global feature.

File: federation/schema.go [hunk 4]
```diff
@@ -196,14 +192,45 @@ func ConvertVersionedSchemas(schemas serviceSchemas) (*SchemaWithFederationInfo,
...
+// single introspection query result that represents the full
+// federated schema.
+func MergeSchemas(schemas serviceSchemas) (*IntrospectionQueryResult, error) {
```
Comment (KevinXing) [inline federation/schema.go:209]: Let's add an unit test for this function?

Comment (KevinXing) [inline federation/schema.go:209]: Maybe rename this function to `MergeIntrospectionSchema`? `MergeSchema` feels merge the full schemas to me.

File: federation/merge_schemas.go (hunk 2 above)
Comment (weilianchu) [reply federation/merge_schemas.go:358]: Yeah, if two directives had the same names, we would only take the first directive. I believe this is the same behaviour as how we merge type fields as well. However, as Zhekai pointed out, the directives of an introspection query are actually hardcoded, so I will be removing this method altogether.

File: federation/merge_schemas.go [hunk 5]
```diff
@@ -312,6 +338,36 @@ func mergeEnumValues(a, b []introspectionEnumValue, mode MergeMode) ([]introspec
...
+	directives := make(map[string][]introspectiveDirective)
+	for _, a := range a {
```
[the remark this answers is not in the document]
Comment (weilianchu) [reply federation/merge_schemas.go:343]: since the directives are hardcoded I will remove this method

**Comment by weilianchu** [comment]
@KevinXing ready for review again

File: federation/schema.go [hunk 6]
```diff
@@ -165,11 +165,7 @@ func validateFieldsReturningFederatedObject(serviceNames []string, serviceSchema
 	return nil
 }
 
-// ConvertVersionedSchemas takes schemas for all of versions of
-// all executors services and generates a single merged schema
-// annotated with mapping from field to all services that know
-// how to resolve the field
-func ConvertVersionedSchemas(schemas serviceSchemas) (*SchemaWithFederationInfo, error) {
+func processSchemaVersions(schemas serviceSchemas) ([]string, []*IntrospectionQueryResult, map[string]*IntrospectionQueryResult, error) {
```
Comment (KevinXing) [inline federation/schema.go:168]: Let's also add a comment for this function to explain the outputs?

File: federation/merge_schemas.go [hunk 7]
```diff
@@ -364,7 +402,43 @@ func mergeTypes(a, b introspectionType, mode MergeMode) (*introspectionType, err
...
+	return directive
+}
+
 func mergeSchemas(a, b *IntrospectionQueryResult, mode MergeMode) (*IntrospectionQueryResult, error) {
+	if a.Schema.QueryType.Name != b.Schema.QueryType.Name {
+		return nil, fmt.Errorf("can't merge QueryType %s and %s", a.Schema.QueryType.Name, b.Schema.QueryType.Name)
+	}
+	if a.Schema.MutationType.Name != b.Schema.MutationType.Name {
+		return nil, fmt.Errorf("can't merge MutationType %s and %s", a.Schema.MutationType.Name, b.Schema.MutationType.Name)
+	}
```
[the remark this answers is not in the document]
Comment (weilianchu) [reply federation/merge_schemas.go:440]: yep, i wrote this for the general case, but if we're specifically dealing with introspection query results, i can hardcode.

**Comment by weilianchu** [comment]
@KevinXing ready for review again

**Review by KevinXing — APPROVED** [review]
Nice!

**Pull request merged**